spring-modulith-events/spring-modulith-events-mongodb/src/test/java/org/springframework/modulith/events/mongodb/MongoDbEventPublicationRepositoryTest.java

```diff
@@ -0,0 +1,206 @@
+class MongoDbEventPublicationRepositoryTest {
+
+	private static final PublicationTargetIdentifier TARGET_IDENTIFIER = PublicationTargetIdentifier.of("listener");
+	private static final String CONNECTION_STRING = "mongodb://%s:%d";
+
+	private static String ip;
+	private static int port;
+	private static MongodExecutable mongodExecutable;
+
+	private MongoTemplate mongoTemplate;
+
+	private MongoDbEventPublicationRepository repository;
+
+	@BeforeAll
+	static void startupMongoDb() throws IOException {
+
+		// Refer to https://www.baeldung.com/spring-boot-embedded-mongodb
+		ip = "localhost";
+		port = Network.freeServerPort(Network.getLocalHost());
+
+		ImmutableMongodConfig mongodConfig = MongodConfig.builder().version(Version.Main.PRODUCTION)
+				.net(new Net(ip, port, Network.localhostIsIPv6())).build();
+
+		MongodStarter starter = MongodStarter.getDefaultInstance();
+		mongodExecutable = starter.prepare(mongodConfig);
+		mongodExecutable.start();
+	}
+
+	@AfterAll
+	static void shutdownMongoDb() {
+		mongodExecutable.stop();
+	}
+
+	@BeforeEach
+	void setUp() {
+		mongoTemplate = new MongoTemplate(MongoClients.create(String.format(CONNECTION_STRING, ip, port)), "test");
+
+		repository = new MongoDbEventPublicationRepository(mongoTemplate);
+	}
+
+	@AfterEach
+	void tearDown() {
+		mongoTemplate.remove(MongoDbEventPublication.class).all();
+	}
+
+	@Test
+	void shouldPersistAndUpdateEventPublication() {
+
+		TestEvent testEvent = new TestEvent("abc");
+
+		CompletableEventPublication publication = CompletableEventPublication.of(testEvent, TARGET_IDENTIFIER);
+
+		// Store publication
+		repository.create(publication);
+
+		List<EventPublication> eventPublications = repository.findIncompletePublications();
+		assertThat(eventPublications).hasSize(1);
+		assertThat(eventPublications.get(0).getEvent()).isEqualTo(publication.getEvent());
+		assertThat(eventPublications.get(0).getTargetIdentifier()).isEqualTo(publication.getTargetIdentifier());
+
+		assertThat(repository.findByEventAndTargetIdentifier(testEvent, TARGET_IDENTIFIER)).isPresent();
+
+		// Complete publication
+		repository.update(publication.markCompleted());
+
+		assertThat(repository.findIncompletePublications()).isEmpty();
+	}
+
+	@Test
+	void shouldUpdateSingleEventPublication() {
+		TestEvent testEvent1 = new TestEvent("id1");
+		TestEvent testEvent2 = new TestEvent("id2");
+
+		CompletableEventPublication publication1 = CompletableEventPublication.of(testEvent1, TARGET_IDENTIFIER);
+		CompletableEventPublication publication2 = CompletableEventPublication.of(testEvent2, TARGET_IDENTIFIER);
+
+		repository.create(publication1);
+		repository.create(publication2);
+
+		repository.update(publication2.markCompleted());
+
+		List<EventPublication> withCompletionDateNull = repository.findIncompletePublications();
+		assertThat(withCompletionDateNull).hasSize(1);
+		assertThat(withCompletionDateNull.get(0).getEvent()).isEqualTo(testEvent1);
+	}
+
+	@Nested
```

Comment:
Do we strictly need this extra level of nesting?

Reply:
We use nested class to group all test cases for one method.